docs/internals/layout_in_memory.rst

```diff
@@ -5,36 +5,121 @@
 
+Reserved Memory Areas
+=====================
+
 Solidity reserves four 32-byte slots, with specific byte ranges (inclusive of endpoints) being used as follows:
 
 - ``0x00`` - ``0x3f`` (64 bytes): scratch space for hashing methods
-- ``0x40`` - ``0x5f`` (32 bytes): currently allocated memory size (aka. free memory pointer)
+- ``0x40`` - ``0x5f`` (32 bytes): allocated memory size (aka. free memory pointer)
 - ``0x60`` - ``0x7f`` (32 bytes): zero slot
 
-Scratch space can be used between statements (i.e. within inline assembly). The zero slot
-is used as initial value for dynamic memory arrays and should never be written to
-(the free memory pointer points to ``0x80`` initially).
+Memory Allocation
+-----------------
+
+Solidity compiles to EVM instructions, and implements its complex data types on top of
+EVM features. Memory allocation is not an EVM feature. Inestead, Solidity implements its
+memory allocation via the free memory pointer stored at location ``0x40``. That value tracks the next
+unallocated memory address available.
+The Solidity compiler generates code that always places new objects at the location pointed to by
+the free memory pointer.
+In inline assembly memory can be allocated manually by increasing the value of the pointer.
+
+Memory located beyond the value of the free memory pointer is considered scratch memory.
+Access to memory in inline assembly is not restricted by the value of the free memory pointer
+and can freely refer to memory located beyond it.
+
+Allocated memory in Solidity is never freed.
+In other words, the value of the free memory pointer is never reduced by Solidity.
+This might change in future versions.
+
+Active Memory and Memory Expansion
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The concept of *allocated memory* is distinct and separate from *active memory*.
+
+The active memory size is tracked by the EVM and represents the highest memory address ever accessed
+by the contract's code.
+Its value is available via the inline assembly instruction ``msize()`` (see
+the :ref:`list of opcodes <opcodes>`).
 
-Solidity always places new objects at the free memory pointer and
-memory is never freed (this might change in the future).
+The size of allocated memory is managed by code automatically generated by the compiler,
+which adjusts the value of the free memory pointer as necessary.
+There are no dedicated opcodes for managing the pointer.
+The slot holding pointer's value is just an ordinary memory location.
 
-Elements in memory arrays in Solidity always occupy multiples of 32 bytes (this
-is even true for ``bytes1[]``, but not for ``bytes`` and ``string``).
+Accessing the free memory pointer itself is a memory access and after such an access ``msize()``
+would (by definition) return a value that is no smaller than the location of the pointer plus its size.
+
+Increasing the size of active memory is called *memory expansion* and incurs an extra gas cost that
+grows non-linearly the size of that memory.
```

Comment:
I'd probably skip mentioning specifics on growth. In fact, the growth is almost linear: for one extra word, it is `(N + 1)^2 // 512 - N^2 // 512 + 3`.  Probably a nitpick.

Reply:
It was actually my suggestion. Originally it only said vaguely that this affects gas.

And, well, O(N^2) seems pretty non-linear to me :)

Reply:
Ah, wait, it's not O(N^2). The subtraction will eliminate the N^2 component...

Reply:
Here's what it said originally: https://github.com/ethereum/solidity/pull/12884#discussion_r840850475